### PR TITLE
Remove Identity op for GemmGradient builder

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: For a model with dynamic shape inputs, an Identity node remains in the final optimized training graph between GemmGrad/ReduceSum node and gradient output. This introduces an extra memcpy from Identity op during training. The solution is to eliminate the node in gradient graph builder. So that final graph could avoid the memcpy and get better performance.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

image

### CODE REVIEW

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -450,8 +450,15 @@ IMPLEMENT_GRADIENT_BUILDER(GetGemmGradient) {
...
         } else {
-          result.push_back(
-              NodeDef("Identity", {IA("dC_reduced")}, {dC}));
+          // skip Identity to avoid memcpy between ReduceSum and graph gradient output
+          NodeDef node_def = result.back();
+          if (node_def.output_args[0] == IA("dC_reduced") && node_def.op_type == "ReduceSum") {
+            node_def.output_args[0] = dC;
+            result.back() = node_def;
+          } else {
+            result.push_back(
```
Comment (ashbhandare) [inline orttraining/orttraining/core/graph/gradient_builder.cc:459]: There are other code paths within GetGemmGradient that add the Identity node. Can we remove them as well?

**Comment by SherlockNoMad** [comment]
In general, Identity op still cames handy when building gradient graph. 
I suggest we fix the graph transformer, and it will fix such case once for all.

**Comment by ytaous** [comment]
thanks for th review, let me cloe this PR and check on EliminateIdentity. :thumbsup: 

---
In reply to: [816827348](https://github.com/microsoft/onnxruntime/pull/7301#issuecomment-816827348) [](ancestors = 816827348)

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 1 above)
Comment (ytaous) [reply orttraining/orttraining/core/graph/gradient_builder.cc:459]: I'll close this PR and investigate EliminateIdentity instead.
Thanks for the review.

---
In reply to: [610763655](https://github.com/microsoft/onnxruntime/pull/7301#discussion_r610763655) [](ancestors = 610763655)

**Comment by ytaous** [comment]
investigate EliminateIdentity instead.

**Pull request closed without merging**